File: design/glassfan-icon/fan_geometry.py
"""Parametric fan blades. One definition feeds the design canvas and, later,
the Swift drawing: radius R=100 at the tips, centre at 0,0, angle 0 straight up,
positive angles clockwise on screen - the direction the app spins the disc."""
import math

def blade_points(r0, sweep, width, tip_blunt, root_frac, sweep_pow=1.5, n=26):
    lead, trail = [], []
    for i in range(n + 1):
        u = i / n
        t = 1 - (1 - u) ** 2                       # denser near the tip
        r = (r0 + t * (1 - r0)) * 100
        phi = sweep * t ** sweep_pow               # centreline angle; negative = tip lags
        hw = width * (1 - t) ** tip_blunt * (root_frac + (1 - root_frac) * 1.35 * t)
        dth = hw / max(r, 1e-6)
        for pts, sign in ((lead, 1), (trail, -1)):
            th = phi + sign * dth
            pts.append((r * math.sin(th), -r * math.cos(th)))
    return trail + lead[::-1][1:]

def smooth_path(pts):
    """Catmull-Rom through the points as cubic Beziers; closed with a straight
    line at the root, which the hub cap hides."""
    f = lambda p: f"{p[0]:.1f},{p[1]:.1f}"
    d = [f"M{f(pts[0])}"]
    n = len(pts)
    for i in range(n - 1):
        p0 = pts[max(i - 1, 0)]; p1 = pts[i]; p2 = pts[i + 1]; p3 = pts[min(i + 2, n - 1)]
        c1 = (p1[0] + (p2[0] - p0[0]) / 6, p1[1] + (p2[1] - p0[1]) / 6)
        c2 = (p2[0] - (p3[0] - p1[0]) / 6, p2[1] - (p3[1] - p1[1]) / 6)
        d.append(f"C{f(c1)} {f(c2)} {f(p2)}")
    d.append("Z")
    return " ".join(d)

DIRECTIONS = {
    # name: count, r0, sweep (rad), width, tip bluntness, root fraction, hub radius
    "swept":    dict(count=6, r0=0.24, sweep=-0.62, width=18, tip_blunt=0.32, root_frac=0.45, hub=27),
    "impeller": dict(count=9, r0=0.36, sweep=-0.95, width=10, tip_blunt=0.30, root_frac=0.55, hub=38, sweep_pow=1.15),
    "four":     dict(count=4, r0=0.20, sweep=0.0,   width=30, tip_blunt=0.42, root_frac=0.28, hub=23),
}

def blade_d(name):
    p = DIRECTIONS[name]
    return smooth_path(blade_points(p["r0"], p["sweep"], p["width"], p["tip_blunt"], p["root_frac"],
                                    p.get("sweep_pow", 1.5)))

if __name__ == "__main__":
    for k in DIRECTIONS:
        print(k, len(blade_d(k)), "chars")

File: Sources/GlassFanUI/FanGeometry.swift
import CoreGraphics
import Foundation

/// The GlassFan blade and the wake behind it - one formula for the dial in the
/// app and for the app icon.
///
/// Designed on the icon canvas as direction A ("six swept"): six crescent blades,
/// tips swept back against the spin. The numbers here are the contract;
/// `Scripts/make-icon.swift` carries a copy of them and
/// `design/glassfan-icon/fan_geometry.py` is where they were drawn.
///
/// Units: radius 1 at the tips, centre at the origin, y pointing down, angle 0
/// straight up and positive angles clockwise on screen - the way the disc spins.
enum FanGeometry {
    static let count = 6
    /// Where the blade leaves the hub, as a share of the tip radius.
    static let root = 0.24
    /// How far the tip lags the root, in radians. Negative: counter-clockwise,
    /// against the spin, so the blade leans back the way its wake trails.
    static let sweep = -0.62
    static let sweepPower = 1.5
    /// Half the blade's width, as arc length at the tip radius.
    static let width = 0.18
    /// Below 1 the tip is rounded rather than pointed.
    static let tipBluntness = 0.32
    /// Width at the root as a share of the widest part.
    static let rootFraction = 0.45
    static let hub = 0.27
    /// The set is turned so a blade's mid-length sits on the vertical axis -
    /// what makes a swept star read as balanced.
    static let orientation = -sweep * 0.55

    /// One blade: the trailing edge from root to tip, then the leading edge back.
    static func outline(samples: Int = 26) -> [CGPoint] {
        var lead: [CGPoint] = [], trail: [CGPoint] = []
        for i in 0...samples {
            let u = Double(i) / Double(samples)
            let t = 1 - (1 - u) * (1 - u)                     // denser towards the tip
            let r = root + t * (1 - root)
            let phi = sweep * pow(t, sweepPower)
            let half = width * pow(1 - t, tipBluntness)
                * (rootFraction + (1 - rootFraction) * 1.35 * t)
            let dth = half / max(r, 1e-6)
            lead.append(point(r, phi + dth))
            trail.append(point(r, phi - dth))
        }
        return trail + lead.reversed().dropFirst()
    }

    private static func point(_ r: Double, _ theta: Double) -> CGPoint {
        CGPoint(x: r * sin(theta), y: -r * cos(theta))
    }

    /// Every blade and the hub, as one path: tips at `radius` around `centre`,
    /// in a y-down space. The outline is smoothed Catmull-Rom; the straight
    /// closing edge at the root sits under the hub.
    static func path(centre: CGPoint, radius: CGFloat, includesHub: Bool = true) -> CGPath {
        let blade = CGMutablePath()
        let pts = outline()
        blade.move(to: pts[0])
        for i in 0..<(pts.count - 1) {
            let p0 = pts[max(i - 1, 0)], p1 = pts[i], p2 = pts[i + 1], p3 = pts[min(i + 2, pts.count - 1)]
            blade.addCurve(to: p2,
                           control1: CGPoint(x: p1.x + (p2.x - p0.x) / 6, y: p1.y + (p2.y - p0.y) / 6),
                           control2: CGPoint(x: p2.x - (p3.x - p1.x) / 6, y: p2.y - (p3.y - p1.y) / 6))
        }
        blade.closeSubpath()

        let all = CGMutablePath()
        for i in 0..<count {
            let turn = CGAffineTransform(translationX: centre.x, y: centre.y)
                .scaledBy(x: radius, y: radius)
                .rotated(by: orientation + Double(i) * 2 * .pi / Double(count))
            all.addPath(blade, transform: turn)
        }
        if includesHub {
            let r = radius * hub
            all.addEllipse(in: CGRect(x: centre.x - r, y: centre.y - r, width: 2 * r, height: 2 * r))
        }
        return all
    }

    /// The wake: copies of the blades turned back against the spin, fading.
    ///
    /// Behind the blade only. The air used to be an angular gradient with a lobe
    /// centred between blades, so it sat on both sides of each one; this is what
    /// a camera sees of anything spinning, and it lies wholly on the trailing
    /// side. Angles are in radians, negative - counter-clockwise, since the disc
    /// turns clockwise. `spread` is how far towards full speed, 0 to 1: nothing
    /// at rest, longer and stronger as the fan speeds up.
    static func wake(spread: Double, ghosts: Int = 16) -> [(angle: Double, opacity: Double)] {
        let s = min(max(spread, 0), 1)
        guard s > 0, ghosts > 1 else { return [] }
        let strength = 0.9 * min(s * 1.6, 1)
        let span = (0.3 + 0.55 * s) * 2 * .pi / Double(count)
        return (1..<ghosts).map { k in
            let f = Double(k) / Double(ghosts)
            return (angle: -f * span, opacity: strength * pow(1 - f, 1.6))
        }
    }
}
